internal/internal_task_pollers.go

```diff
@@ -253,6 +260,90 @@ func (wtp *workflowTaskPoller) processResetStickinessTask(rst *resetStickinessTa
 	return err
 }
 
+func (wtp *workflowTaskPoller) scheduleRespondDecisionTaskCompleted(wc *WorkflowExecutionContext, workflowTask *workflowTask, startTime time.Time) {
+	deadlineToTrigger := time.Duration(float32(0.8) * float32(wc.workflowInfo.TaskStartToCloseTimeoutSeconds) * float32(time.Second))
+	delayDuration := startTime.Add(deadlineToTrigger).Sub(time.Now())
+	time.AfterFunc(delayDuration, func() {
+		wtp.forceRespondDecisionTaskCompleted(wc, workflowTask, startTime)
+	})
+}
+
+func (wtp *workflowTaskPoller) forceRespondDecisionTaskCompleted(wc *WorkflowExecutionContext, workflowTask *workflowTask, startTime time.Time) {
+	wc.Lock()
+	defer wc.release()
+
+	if wc.currentDecisionTask != workflowTask.task {
+		// decision task already completed
+		var currentTaskStartedEventID int64
+		if wc.currentDecisionTask != nil {
+			currentTaskStartedEventID = wc.currentDecisionTask.GetStartedEventId()
+		}
+		wtp.logger.Debug("DecisionTask already completed when force responding timer fires.",
+			zap.Int64("TaskStartedEventID", workflowTask.task.GetStartedEventId()),
+			zap.Int64("CurrentStartedEventID", currentTaskStartedEventID))
+		return
+	}
+
+	completeRequest := wc.CompleteDecisionTaskWithLock(true)
+	wtp.logger.Debug("Force RespondDecisionTaskCompleted.",
+		zap.Int64("TaskStartedEventID", workflowTask.task.GetStartedEventId()))
+	wtp.metricsScope.Counter(metrics.DecisionTaskForceCompleted).Inc(1)
+
+	response, err := wtp.RespondTaskCompletedWithMetrics(completeRequest, workflowTask.task, startTime)
+	if err != nil {
+		return
+	}
+
+	wc.processCompleteDecisionResponseWithLock(response, wtp)
+	return
+}
+
+func (w *WorkflowExecutionContext) processCompleteDecisionResponseWithLock(response *s.RespondDecisionTaskCompletedResponse, wtp *workflowTaskPoller) error {
```

Comment:
At google they had standard to call such functions ....Locked. It would make it processCompleteDecisionResponseLocked

Reply:
good to know, will update